Rewrite List doc (#6970)
marmelab/react-admin: docs/ListTutorial.md

                    <TextField source="title" />
                    <TextField source="author" />
                    <TextField source="year" />
                </Datagrid>
            </Card>
            <Toolbar>
                {page > 1 && <Button onClick={() => setPage(page - 1)}>Previous page</Button>}
                {page < total / perPage && <Button onClick={() => setPage(page + 1)}>Next page</Button>}
            </Toolbar>
        </div>
    );
};
```

`<Datagrid>`does more than the previous table: it renders table headers depending on the current sort, and allows you to change the sort order by clicking a column header. Also, for each row, `<Datagrid>` creates a `RecordContext`, which lets you use react-admin Field and Buttons without explicitely passing the row data.

## `ListContext` Exposes List Data To Descendents

`<Datagrid>` requires a `data` prop to render, but it can grab it from a `ListContext` instead. Creating such a context with `<ListContextProvider>` also allows to use other react-admin components specialized in filtering (`<FilterForm>`) and pagination (`<Pagination>`), and to reduce the boilerplate code even further:

{% raw %}

> ```jsx

Reviewer: ```suggestion
```jsx
import * as React from 'react';
```

Author: This is intended

As calling the List controller and putting its result into a context is also common, react-admin provides [the `<ListBase>` component](./ListBase.md) to do it. So the example can be further simplified to the following: 

```jsx
import { 
    Title,
    ListBase,
    Datagrid,
    TextField,
    FilterForm,
    Pagination,
    TextInput
} from 'react-admin';
import { Card } from '@mui/material';

const filters = [<TextInput label="Search" source="q" size="small" alwaysOn />];

const BookList = () => (
    <ListBase>
        <div>
            <Title title="Book list" />
            <FilterForm filters={filters} />
            <Card>
                <Datagrid>
                    <TextField source="id" />
                    <TextField source="title" />
                    <TextField source="author" />
                    <TextField source="year" />
                </Datagrid>
            </Card>
            <Pagination />
        </div>
    </ListBase>
);
```

Notice that we're not handling the loading state manually anymore. In fact, the `<Datagrid>` component can render a skeleton while the data is being fetched.

## `useListContext` Accesses The List Context

> Using the `<ListBase>` component has one drawback: you can no longer access the list context (`data`, `total`, etc) in the component. Instead, you have to access it from the `ListContext` using [the `useListContext` hook](./useListContext.md).

Reviewer: ```suggestion
Using the `<ListBase>` component has one drawback: you can no longer access the list context (`data`, `total`, etc.) in the component. Instead, you have to access it from the `ListContext` using [the `useListContext` hook](./useListContext.md).
```

```jsx
import { Admin, Resource, ListGuesser } from 'react-admin';

const App = () => (
    <Admin dataProvider={...}>
        <Resource name="posts" list={ListGuesser} />
    </Admin>
);
```

`<ListGuesser>` is also a good way to bootstrap a List view, as it outputs the code that it generated for the List into the console. Copy and paste that code in a custom List component, and you can start customizing the list view in no time.

## Using Another Layout

On Mobile, `<Datagrid>` doesn't work well - the screen is too narrow. You can use the `<SimpleList>` component instead. It's a drop-in replacement for `<Datagrid>`.

```jsx
import { 
    List,
    SimpleList,
    TextInput
} from 'react-admin';
import { Card } from '@mui/material';

const filters = [<TextInput label="Search" source="q" size="small" alwaysOn />];

const BookList = () => (
    <List filters={filters}>
        <SimpleList 
            primaryText={record => <i>record.title</i>}
            secondaryText={record => <>By {record.author} ({record.year})</>}
        />
    </List>
);
```

> React-admin offers many more List layouts: `<EditableDatagrid>`, `<TreeWithDetails>`, `<Calendar>`, etc. The `<List>` component accepts any component as child - even components of your own! 

Reviewer 2: ```suggestion
React-admin offers many more List layouts: [`<EditableDatagrid>`](./EditableDatagrid.md), [`<TreeWithDetails>`](./Tree.md), [`<Calendar>`](./Calendar.md), etc. The `<List>` component accepts any component as child - even components of your own! 
```

For these cases, react-admin provides a `<ListGuesser>` component that will guess the datagrid columns from the data. It's a bit like the `<List>` component, but it doesn't require any configuration.

```jsx
import { Admin, Resource, ListGuesser } from 'react-admin';

const App = () => (
    <Admin dataProvider={...}>
        <Resource name="posts" list={ListGuesser} />
    </Admin>
);
```

`<ListGuesser>` is also a good way to bootstrap a List view, as it outputs the code that it generated for the List into the console. Copy and paste that code in a custom List component, and you can start customizing the list view in no time.

## List Iterators

The components you can use as child of `<List>` are called "List Iterator". They render a list of records. `<Datagrid>` is such a List Iterator, but react-admin provides many more:

- [`<Datagrid>`](./Datagrid.md)
- [`<SimpleList>`](./SimpleList.md)
- [`<SingleFieldList>`](./SingleFieldList.md)
- [`<EditableDatagrid>`](./EditableDatagrid.md)

> - [`<TreeWithDetails>`](./Tree.md)

Reviewer: ```suggestion
- [`<TreeWithDetails>`](./TreeWithDetails.md)
```

## Filtering the List

One of the most important features of the List page is the ability to search for a dedicated record. In this documentation, we use the term "filter" for the cpontrols allowing to search the list.

React-admin rovides 2 possible UIs for filters, and lets you you your own if they're not sufficient.

The first filter UI is called "the Filer/Form Combo". 

![Filter Button/Form Combo](./img/list_filter.gif)

The second filter UI is called "the Filter List Sidebar".

![Filter Button/Form Combo](./img/filter-sidebar.gif)

Check [the dedicated Filter tutorial chapter](./FilteringTutorial.md) for more information on filtering.

## Sorting the List

The List view uses the `sort` and `order` query parameters to determine the sort field and order passed to `dataProvider.getList()`.

Here is a typical List URL:

> https://myadmin.dev/#/posts?displayedFilters=%7B%22commentable%22%3Atrue%7D&filter=%7B%22commentable%22%3Atrue%2C%22q%22%3A%22lorem%20%22%7D&order=DESC&page=1&perPage=10&sort=published_at

Once decoded, this URL reveals the intended sort:

```
sort=published_at
order=DESC
```

> If you're using a `<Datagrid>` inside the List view, then the column headers are buttons allowing users to change the list sort field and order. This feature requires no configuration and works out fo the box. Check [the `<Datagrid>` documentation](./Datagrid.md#customizing-the-sort-order-for-columns) to see how to disable or modify the field used for sorting on a particular column.

Reviewer: ```suggestion
If you're using a `<Datagrid>` inside the List view, then the column headers, are buttons allowing users to change the list sort field and order. This feature requires no configuration and works out of the box. Check [the `<Datagrid>` documentation](./Datagrid.md#customizing-column-sort) to see how to disable or modify the field used for sorting on a particular column.
```
